Replace GetService with GetServiceAsync calls in Visual Studio

erdembayar opened the pull request:
Fixes: https://github.com/NuGet/Home/issues/3784

Regression? No
## Description
Here addressing "Recommended practice of acquiring VS services is always calling GetServiceAsync to avoid the high risk of deadlock". It's incremental effort and this is part 2. Here reduced sync calls for all remaining GetService calls not addressed in [PR#3766](https://github.com/NuGet/NuGet.Client/pull/3766).

Replaced non-async with async calls in following files:
src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourceOptionsPage.cs
src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs
src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VsCommonOperations.cs
src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/OptionsPageActivator.cs
src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VSSourceControlManagerProvider.cs
src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/TelemetryUtility.cs
src/NuGet.Clients/NuGet.VisualStudio.Implementation/VsTemplateWizard.cs

Removed dead(no longer relevant) codes from following files due to above refactoring:
src/NuGet.Clients/NuGet.VisualStudio.Common/GlobalSuppressions.cs
src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceProviderExtensions.cs
src/NuGet.Clients/NuGet.VisualStudio.Implementation/GlobalSuppressions.cs

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [x] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

erdembayar committed:
Replace .GetService calls with .GetServiceAsync calls.

erdembayar committed:
Add remaining .GetService to .GetServiceAsync replacement.

zivkan commented:
This PR has the string `.Run(` 11 times.  Are you sure this achieves anything in relation to reducing risk of deadlock? JTF will only allow async tasks "related" to the delegate/action passed to `Run`. Any other async task, or sync code, that is trying to switch to the UI thread is still going to be blocked until `.Run(...)` returns.

Also, can you post a link to the source of the recommended practice? Since you put it in double quotes, I assumed you copy/pasted it from somewhere, but searching for it online only find your own PR. I'm curious if the recommendation gives more details.

Secondly, in a few places you removed `IServiceProvider` or `IAsyncServiceProvider` being passed in method calls, and instead use a global/static async service provider. This makes the code less testable, as it's no longer possible to easily pass in a mock service provider that resolves mock services. Even if the static service provider allows tests to override it, it means that tests can't be run in parallel. Even though the classes you modified aren't currently being tested, so it "doesn't matter", it raises the barrier for anyone who wants to start testing the class in the future, so I don't think it's a good change.

nkolev92 commented on src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs:
```diff
@@ -51,7 +52,15 @@ private PowerConsoleWindow PowerConsoleWindow
-            get { return this.GetService<IVsUIShell>(typeof(SVsUIShell)); }
+            get
+            {
+                return NuGetUIThreadHelper.JoinableTaskFactory.Run(GetIVsUIShellAsync);
```
I think there's a better way to do this. 

The `GetService` call that used to be here is a consequence of this getting invoked on a synchronous codepath.

Changing to JTF.Run remove the GetService call, but doesn't really do anything in the grandscheme of improving perf. 

To improve perf, you should go another layer up. 

For example, this is *only* in https://github.com/NuGet/NuGet.Client/blob/a1fa3375d91d95e617510480f4c3b28324563f94/src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs#L521. 

Which is already an async codepath, so the *correct* thing to do here would be to call GetServiceAsync *directly*. 

Same applies to *all* situations.

erdembayar commented:
> Also, can you post a link to the source of the recommended practice? Since you put it in double quotes, I assumed you copy/pasted it from somewhere, but searching for it online only find your own PR. I'm curious if the recommendation gives more details.
> 

It's just body of initial issue. https://github.com/NuGet/Home/issues/3784#issue-185520889
Actually I don't have any good guide, I'm just focusing on replace all '.GetReplace' vs '.GetReplaceAsync' which are listed by @dominoFire [here](https://github.com/NuGet/Home/issues/3784#issue-185520889).

nkolev92 reviewed (commented):
I think we haven't clarified the expectations for this task too well. 

I left a comment that hopefully improves that. 
I'd be happy to have an offline chat tomorrow if need be.

erdembayar committed:
Remove NuGetUIThreadHelper.JoinableTaskFactory.Run by implementing all the way up async call.

erdembayar committed:
Revert some changes after code review since they're not worth changing to async version because still run synch.

erdembayar committed:
After consulting with Jean-Pierre it's ok to use async with event handler.

erdembayar committed:
Remove GetDTE non-async method and use actual async GetDTEAsync

erdembayar committed:
Convert PackageSourceOptionsPage to async

erdembayar committed:
Revert not benefitial change.

erdembayar committed:
Make OptionsPageActivator async

erdembayar committed:
Git revert unneeded changes.

erdembayar committed:
Revert some unneeded change.

erdembayar committed:
Add exception telemetry for async calls.

erdembayar commented:
@NuGet/nuget-client 
Ok. I addressed PR comment concerns and I tried to make it async all the way up.
Also I introduced some async event delegate handlers based on existing pattern we already have:
https://github.com/NuGet/NuGet.Client/blob/f754a0da911a07cfac987df0fbdf5518692e3d5a/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs#L41-L49

Almost all 3 cases they're working on UI so I don't believe they introduce any timing bug since they're already fire and forget type things.

1. PowerConsoleToolWindow.cs : Update menu and commands of VS. 
2. PackageSourcesOptionsControl.cs : Open browse explorer in "Package Source Option"
![image](https://user-images.githubusercontent.com/8766776/108568389-de11e700-72be-11eb-911a-52d417592cd6.png)

3. OptionsPageActivator.cs : Open "General" and "Package sources"
![image](https://user-images.githubusercontent.com/8766776/108568313-bfabeb80-72be-11eb-9237-37afdb839205.png)

![image](https://user-images.githubusercontent.com/8766776/108568245-95f2c480-72be-11eb-86ae-09a927b8b35a.png)

erdembayar commented:
> This PR has the string `.Run(` 11 times. Are you sure this achieves anything in relation to reducing risk of deadlock? JTF will only allow async tasks "related" to the delegate/action passed to `Run`. Any other async task, or sync code, that is trying to switch to the UI thread is still going to be blocked until `.Run(...)` returns.
> 
Thank you for your review.

Taken your concern in consideration. Unless at least one direct async call benefit I'm not adding async wrapped in `.Run()` by default.

> Also, can you post a link to the source of the recommended practice? Since you put it in double quotes, I assumed you copy/pasted it from somewhere, but searching for it online only find your own PR. I'm curious if the recommendation gives more details.
> 
Sorry, I don't have recommended practice. That was just body text of this [issue](https://github.com/NuGet/Home/issues/3784#issue-185520889).

> Secondly, in a few places you removed `IServiceProvider` or `IAsyncServiceProvider` being passed in method calls, and instead use a global/static async service provider. This makes the code less testable, as it's no longer possible to easily pass in a mock service provider that resolves mock services. Even if the static service provider allows tests to override it, it means that tests can't be run in parallel. Even though the classes you modified aren't currently being tested, so it "doesn't matter", it raises the barrier for anyone who wants to start testing the class in the future, so I don't think it's a good change.

Taken your advise and replacing constructor based service provider with global service provider. But I did replace non-async service provider with async service provider through constructor injection.

nkolev92 commented on src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs:
```diff
@@ -49,9 +51,9 @@ private PowerConsoleWindow PowerConsoleWindow
         }
 
-        private IVsUIShell VsUIShell
+        private async Task<IVsUIShell> GetIVsUIShellAsync()
```
Given that this is called in 1 place only, you don't need the extra allocation. Just make the async provider call in the calling method instead.

erdembayar replied:
Ok. Addressed.

nkolev92 commented on src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs:
```diff
@@ -472,15 +474,19 @@ private void StartConsoleSession(FrameworkElement consolePane)
                     if (WpfConsole.Dispatcher.IsStartCompleted)
                     {
-                        OnDispatcherStartCompleted();
+                        NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
```
You don't really need the extra delegate: 

```cs
                        NuGetUIThreadHelper.JoinableTaskFactory.Run(OnDispatcherStartCompletedAsync);
``` 

The above should do the job as well.

erdembayar replied:
Ok. Addressed.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs:
```diff
@@ -607,7 +609,15 @@ private void UpdateTextBoxes(PackageSourceContextInfo packageSource)
-        private void OnBrowseButtonClicked(object sender, EventArgs e)
+        private void OnBrowseButtonClicked(object sender, EventArgs args)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async delegate
```
You don't need an extra delegate here. 
See above comment.

erdembayar replied:
Ok. Addressed.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs:
```diff
@@ -607,7 +609,15 @@ private void UpdateTextBoxes(PackageSourceContextInfo packageSource)
+
+        private async Task OnBrowseButtonClickedAsync()
         {
             ThreadHelper.ThrowIfNotOnUIThread();
```
Given that this is now async, this check is incorrect. 

See rule #1 https://github.com/microsoft/vs-threading/blob/main/doc/threading_rules.md

erdembayar replied:
Ok. Replaced with `await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();`

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VsCommonOperations.cs:
```diff
@@ -14,43 +15,33 @@ namespace NuGet.PackageManagement.VisualStudio
-                throw new ArgumentNullException(nameof(serviceProvider));
-            }
-
-            _dte = new Lazy<EnvDTE.DTE>(
```
Given the scenario, It might not matter too much, but why did you remove the caching of the DTE object here?

erdembayar replied:
Sorry, not intentional. Replaced with AsyncLazy, please check if implementation is correct.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/OptionsPageActivator.cs:
```diff
@@ -54,25 +53,32 @@ public void ActivatePage(OptionsPage page, Action closeCallback)
-        private void ShowOptionsPage(string optionsPageGuid)
+        private async Task ShowOptionsPageAsync(string optionsPageGuid)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
```
VS threading rule #1.

erdembayar replied:
Ok. Replaced with `await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();`

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VSSourceControlManagerProvider.cs:
```diff
@@ -15,41 +15,39 @@ namespace NuGet.PackageManagement.VisualStudio
 
             _settings = vsSettings;
-
-            _dte = new Lazy<EnvDTE.DTE>(
```
Was the removal of the DTE caching intentional?

erdembayar replied:
Sorry, not intentional. Replaced with AsyncLazy, please check if implementation is correct.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/VsTemplateWizard.cs:
```diff
@@ -284,15 +284,15 @@ private async Task TemplateFinishedGeneratingAsync(Project project)
-        private void RunDesignTimeBuild(Project project)
+        private async Task RunDesignTimeBuildAsync(Project project)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
```
vs threading rule #1.

erdembayar replied:
ok. Replaced with `await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();`

nkolev92 reviewed (commented):
Looks a lot better. 

A couple of VS rule violations to be addressed.

erdembayar committed:
Address PR review by Nikolche.

erdembayar reviewed (commented):
@nkolev92 
Thank you for your review. I addressed your comment please check.

nkolev92 commented on src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs:
```diff
@@ -507,19 +505,27 @@ private void OnDispatcherStartWaitingKey(object sender, EventArgs args)
+        {
+            WpfConsole.Dispatcher.StartCompleted -= OnDispatcherStartCompleted;
+
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async delegate
```
You have an extra delegate here.

erdembayar committed:
Address Nikolche's review about delegate.

erdembayar commented:
@NuGet/nuget-client 
Please review. Need 1 more review  approval.
